--recursive fail to parse folder with __init__.py at the root when using relative path

### Bug description

```python
# folder struct
./__init__.py
./a.py # same goes if a.py is in a subfolder

# __init__.py is empty

# a.py includes anything that may trigger an error report like:

def foo():
    unused = 1
```


### Configuration

_No response_

### Command used

```shell
pylint --recursive=y ./
```


### Pylint output

```python
None
```


### Expected behavior

Should be like `pylint a.py` or `pylint --recursive=y $(pwd)`

```
************* Module testpy.a
a.py:1:0: C0114: Missing module docstring (missing-module-docstring)
a.py:1:0: C0116: Missing function or method docstring (missing-function-docstring)
a.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
a.py:2:4: W0612: Unused variable 'unused' (unused-variable)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Pylint version

```shell
pylint 3.0.2
astroid 3.0.1
Python 3.11.6 (main, Oct  2 2023, 20:46:17) [Clang 14.0.0 (clang-1400.0.29.202)]
```


### OS / Environment

MacOS

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/9210.bugfix
+++ b/doc/whatsnew/fragments/9210.bugfix
@@ -1,4 +1,4 @@
-Fix bug for not being able to walk through files when work on `./` as input
-at a directory with a `__init__.py` file.
+Fix a bug where pylint was unable to walk recursively through a directory if the
+directory has an `__init__.py` file.
 
 Closes #9210

--- a/doc/whatsnew/fragments/9210.bugfix
+++ b/doc/whatsnew/fragments/9210.bugfix
@@ -1,0 +1,4 @@
+Fix bug for not being able to walk through files when work on `./` as input
+at a directory with a `__init__.py` file.
+
+Closes #9210

--- a/pylint/lint/expand_modules.py
+++ b/pylint/lint/expand_modules.py
@@ -144,8 +144,9 @@ def expand_modules(
         )
         if has_init or is_namespace or is_directory:
             for subfilepath in modutils.get_module_files(
-                os.path.dirname(filepath), ignore_list, list_all=is_namespace
+                os.path.dirname(filepath) or ".", ignore_list, list_all=is_namespace
             ):
+                subfilepath = os.path.normpath(subfilepath)
                 if filepath == subfilepath:
                     continue
                 if _is_in_ignore_list_re(

--- a/tests/lint/unittest_expand_modules.py
+++ b/tests/lint/unittest_expand_modules.py
@@ -7,9 +7,9 @@ from __future__ import annotations
 import copy
 import os
 import re
+from collections.abc import Iterator
 from contextlib import contextmanager
 from pathlib import Path
-from typing import Iterator
 
 import pytest
 

--- a/tests/lint/unittest_expand_modules.py
+++ b/tests/lint/unittest_expand_modules.py
@@ -4,11 +4,11 @@
 
 from __future__ import annotations
 
-from contextlib import contextmanager
 import copy
 import os
-from pathlib import Path
 import re
+from contextlib import contextmanager
+from pathlib import Path
 from typing import Iterator
 
 import pytest
@@ -134,13 +134,15 @@ def _list_expected_package_modules_relative() -> tuple[dict[str, object], ...]:
     """Generates reusable list of modules for our package with relative path input."""
     abs_result = copy.deepcopy(_list_expected_package_modules())
     for item in abs_result:
+        assert isinstance(item["basepath"], str)
+        assert isinstance(item["path"], str)
         item["basepath"] = os.path.relpath(item["basepath"],  str(Path(__file__).parent))
         item["path"] = os.path.relpath(item["path"],  str(Path(__file__).parent))
     return abs_result
 
 
 @contextmanager
-def pushd(path: Path) -> Iterator[str]:
+def pushd(path: Path) -> Iterator[None]:
     prev = os.getcwd()
     os.chdir(path)
     try:
@@ -194,11 +196,12 @@ class TestExpandModules(CheckerTestCase):
     @pytest.mark.parametrize(
         "files_or_modules,expected",
         [
-            ([Path(__file__).name], {
-                this_file_relative_to_parent["path"]: this_file_relative_to_parent
-            }),
             (
-                ['./'],
+                [Path(__file__).name],
+                {this_file_relative_to_parent["path"]: this_file_relative_to_parent},
+            ),
+            (
+                ["./"],
                 {
                     module["path"]: module  # pylint: disable=unsubscriptable-object
                     for module in _list_expected_package_modules_relative()

--- a/tests/lint/unittest_expand_modules.py
+++ b/tests/lint/unittest_expand_modules.py
@@ -4,8 +4,12 @@
 
 from __future__ import annotations
 
-import re
+from contextlib import contextmanager
+import copy
+import os
 from pathlib import Path
+import re
+from typing import Iterator
 
 import pytest
 
@@ -28,13 +32,22 @@ def test__is_in_ignore_list_re_match() -> None:
 
 TEST_DIRECTORY = Path(__file__).parent.parent
 INIT_PATH = str(TEST_DIRECTORY / "lint/__init__.py")
-EXPAND_MODULES = str(TEST_DIRECTORY / "lint/unittest_expand_modules.py")
+EXPAND_MODULES_BASE = "unittest_expand_modules.py"
+EXPAND_MODULES = str(TEST_DIRECTORY / "lint" / EXPAND_MODULES_BASE)
 this_file = {
     "basename": "lint.unittest_expand_modules",
     "basepath": EXPAND_MODULES,
     "isarg": True,
     "name": "lint.unittest_expand_modules",
     "path": EXPAND_MODULES,
+}
+
+this_file_relative_to_parent = {
+    "basename": "lint.unittest_expand_modules",
+    "basepath": EXPAND_MODULES_BASE,
+    "isarg": True,
+    "name": "lint.unittest_expand_modules",
+    "path": EXPAND_MODULES_BASE,
 }
 
 this_file_from_init = {
@@ -117,6 +130,25 @@ def _list_expected_package_modules(
     )
 
 
+def _list_expected_package_modules_relative() -> tuple[dict[str, object], ...]:
+    """Generates reusable list of modules for our package with relative path input."""
+    abs_result = copy.deepcopy(_list_expected_package_modules())
+    for item in abs_result:
+        item["basepath"] = os.path.relpath(item["basepath"],  str(Path(__file__).parent))
+        item["path"] = os.path.relpath(item["path"],  str(Path(__file__).parent))
+    return abs_result
+
+
+@contextmanager
+def pushd(path: Path) -> Iterator[str]:
+    prev = os.getcwd()
+    os.chdir(path)
+    try:
+        yield
+    finally:
+        os.chdir(prev)
+
+
 class TestExpandModules(CheckerTestCase):
     """Test the expand_modules function while allowing options to be set."""
 
@@ -158,6 +190,41 @@ class TestExpandModules(CheckerTestCase):
         )
         assert modules == expected
         assert not errors
+
+    @pytest.mark.parametrize(
+        "files_or_modules,expected",
+        [
+            ([Path(__file__).name], {
+                this_file_relative_to_parent["path"]: this_file_relative_to_parent
+            }),
+            (
+                ['./'],
+                {
+                    module["path"]: module  # pylint: disable=unsubscriptable-object
+                    for module in _list_expected_package_modules_relative()
+                },
+            ),
+        ],
+    )
+    @set_config(ignore_paths="")
+    def test_expand_modules_relative_path(
+        self, files_or_modules: list[str], expected: dict[str, ModuleDescriptionDict]
+    ) -> None:
+        """Test expand_modules with the default value of ignore-paths and relative path as input."""
+        ignore_list: list[str] = []
+        ignore_list_re: list[re.Pattern[str]] = []
+        with pushd(Path(__file__).parent):
+            modules, errors = expand_modules(
+                files_or_modules,
+                [],
+                ignore_list,
+                ignore_list_re,
+                self.linter.config.ignore_paths,
+            )
+            print(modules)
+            print(expected)
+            assert modules == expected
+            assert not errors
 
     @pytest.mark.parametrize(
         "files_or_modules,expected",

--- a/tests/lint/unittest_expand_modules.py
+++ b/tests/lint/unittest_expand_modules.py
@@ -136,8 +136,8 @@ def _list_expected_package_modules_relative() -> tuple[dict[str, object], ...]:
     for item in abs_result:
         assert isinstance(item["basepath"], str)
         assert isinstance(item["path"], str)
-        item["basepath"] = os.path.relpath(item["basepath"],  str(Path(__file__).parent))
-        item["path"] = os.path.relpath(item["path"],  str(Path(__file__).parent))
+        item["basepath"] = os.path.relpath(item["basepath"], str(Path(__file__).parent))
+        item["path"] = os.path.relpath(item["path"], str(Path(__file__).parent))
     return abs_result
 
 

--- a/tests/lint/unittest_expand_modules.py
+++ b/tests/lint/unittest_expand_modules.py
@@ -224,8 +224,6 @@ class TestExpandModules(CheckerTestCase):
                 ignore_list_re,
                 self.linter.config.ignore_paths,
             )
-            print(modules)
-            print(expected)
             assert modules == expected
             assert not errors
 

